Construct multi-qubit Pauli terms and sums from strings

jlbosse opened the pull request:
Description
-----------

Added `PauliSum.from_compact_str()` and `PauliSum.compact_str()` in analogy to `PauliTerm.from_compact_str()` and `PaulTerm.compact_str()`. 
Added support for multi-qubit operators to `PauliTerm.from_compact_str()`.
Added corresponding tests.

This resolves #977 mostly (except for the question, whether __init__() should also be changed).

Checklist
---------

- [x] The above description motivates these changes.
- [x] There is a unit test that covers these changes.
- [x] All new and existing tests pass locally and on Semaphore.
  test_examples.test_quantum_die fails, but since I ddin't touch anything remotely related I
  am sure I can shift the blame
- [x] Parameters have type hints with [PEP 484 syntax](https://www.python.org/dev/peps/pep-0484/).
  None of the surrounding code has type hints, so I refrained from adding them.
- [x] Functions and classes have useful sphinx-style docstrings.
- [x] (New Feature) The docs have been updated accordingly.
- [x] (Bugfix) The associated issue is referenced above using
      [auto-close keywords](https://help.github.com/en/articles/closing-issues-using-keywords).
- [x] The [changelog](https://github.com/rigetti/pyquil/blob/master/CHANGELOG.md) is updated,
      including author and PR number (@username, gh-xxx).

appleby commented on pyquil/paulis.py:
```diff
@@ -677,6 +684,28 @@ def get_programs(self):
+        """A string representation of the PauliSum that is more compact than ``str(pauli_sum)``
+
+        >>> pauli_sum = 2.0 * sX(1)* sZ(2) + 1.5 * sY(2)
+        >>> str(term)
```
`term` -> `pauli_sum` ? ditto just below

jlbosse replied:
Oops, fixed now.

karalekas commented:
As for `__init__`, let's leave that question for another PR. 👍

jlbosse commented:
@karalekas Thanks for pointing out the manifold ways the user could specify unparsable strings. I added checks for each part of the string now that point at the exact part where things became unparsable.

 - For the case `"(1.0+0j)*X0 + (1.0+0j)*Y0 + 1*X0"` I added a simple `.simplify()` before return.
 - `"(1.0+0j)*A0"` and `"X0 + X1"` now throw more informative exceptions.
 - `"1+1j*X0"` is still considered equivalent to `"(1+1j)*X0"`. Should I add another check, that '+' only shows up inside parenthesis when part of a `pauli_term_str`? Note, that this is only potentially confusing when passed to `PauliTerm.from_compact_str()`. `PauliSum.from_compact_str()`, on the other hand,  will choke on `"1+1j*X0"` and raise a value error telling you what went wrong.

jlbosse commented:
Done! I added tests for the simplification part, strings containing the identity and sums of length one.

karalekas commented on CHANGELOG.md:
```diff
@@ -77,6 +77,9 @@ Changelog
 -   PyQuil now sends “modern” ISA payloads to quilc, which must be of version
     \>= `1.10.0`. Check out the details of `get_isa` for information on how to
     specify custom payloads (@ecpeterson, gh-961).
+-   PauliSums can now be constructed from strings and
```
```suggestion
-   `PauliSum` objects can now be constructed from strings via `from_compact_str()`
```

karalekas commented on CHANGELOG.md:
```diff
@@ -77,6 +77,9 @@ Changelog
     \>= `1.10.0`. Check out the details of `get_isa` for information on how to
     specify custom payloads (@ecpeterson, gh-961).
+-   PauliSums can now be constructed from strings and
+    PauliTerm.from_compact_str() supports multi qubit strings. Added
```
```suggestion
    and `PauliTerm.from_compact_str()` supports multi-qubit strings (@jlbosse, gh-984).
```

karalekas commented on CHANGELOG.md:
```diff
@@ -77,6 +77,9 @@ Changelog
     specify custom payloads (@ecpeterson, gh-961).
+-   PauliSums can now be constructed from strings and
+    PauliTerm.from_compact_str() supports multi qubit strings. Added
+    PauliSum.compact_str() in analogy to PauliTerm.compact_str(). (@jlbosse, gh-984)
```
```suggestion
```

jlbosse commented:
Incorporated the Changelog changes (and clearly didn't write that fact into the changelog, otherwise I would still be stuck in a recursive loop without exit condition). Rebasing off master has been done commit e761981.

karalekas commented:
@jlbosse Congrats on your first PR contribution to pyQuil! :)
